components/camera/client.go

```diff
@@ -356,143 +365,204 @@ func (c *client) Close(ctx context.Context) error {
-	if err := c.unsubscribeAll(); err != nil {
-		c.logger.CErrorw(ctx, "Close > unsubscribeAll", "err", err, "name", c.Name())
-		return err
-	}
+	c.unsubscribeAll()
 
+	// NOTE: (Nick S) we are intentionally releasing the lock before we wait for
+	// background goroutines to terminate as some of them need to be able
+	// to take the lock to terminate
+	c.mu.Unlock()
+	c.activeBackgroundWorkers.Wait()
 	return nil
 }
 
+// SubscribeRTP begins a subscription to receive RTP packets.
+// When the Subscription terminates the context in the returned Subscription
+// is cancelled.
+// It maintains the invariant that there is at most a single track between
+// the client and WebRTC peer.
+// It is strongly recommended to set a timeout on ctx as the underlying
+// WebRTC peer connection's Track can be removed before sending an RTP packet which
+// results in SubscribeRTP blocking until the provided context is cancelled or the client
+// is closed.
+// This rare condition may happen in cases like reconfiguring concurrently with
+// a SubscribeRTP call.
 func (c *client) SubscribeRTP(
 	ctx context.Context,
 	bufferSize int,
 	packetsCB rtppassthrough.PacketCallback,
-) (rtppassthrough.SubscriptionID, error) {
+) (rtppassthrough.Subscription, error) {
 	ctx, span := trace.StartSpan(ctx, "camera::client::SubscribeRTP")
 	defer span.End()
 	c.mu.Lock()
 	defer c.mu.Unlock()
-
-	onError := func(err error) { c.logger.CErrorw(ctx, "StreamSubscription returned hit an error", "err", err) }
-	sub, err := rtppassthrough.NewStreamSubscription(bufferSize, onError)
+	sub, err := rtppassthrough.NewStreamSubscription(bufferSize)
 	if err != nil {
-		return uuid.Nil, err
+		return rtppassthrough.NilSubscription, err
 	}
 
 	if c.conn.PeerConn() == nil {
-		return uuid.Nil, ErrNoPeerConnection
+		return rtppassthrough.NilSubscription, ErrNoPeerConnection
 	}
 
 	// check if we have established a connection that can be shared by multiple clients asking for cameras streams from viam server.
 	sc, ok := c.conn.(*rdkgrpc.SharedConn)
 	if !ok {
-		return uuid.Nil, ErrNoSharedPeerConnection
+		return rtppassthrough.NilSubscription, ErrNoSharedPeerConnection
 	}
 	c.logger.CDebugw(ctx, "SubscribeRTP", "subID", sub.ID().String(), "name", c.Name(), "subAndCallbackByID", c.subAndCallbackByID.String())
 	defer func() {
 		c.logger.CDebugw(ctx, "SubscribeRTP after", "subID", sub.ID().String(),
 			"name", c.Name(), "subAndCallbackByID", c.subAndCallbackByID.String())
 	}()
 
-	// add the subscription to subAndCallbackByID so the goroutine spawned by
-	// addOnTrackSubFunc can forward the packets it receives from the modular camera
-	// over WebRTC to the SubscribeRTP caller via the packetsCB callback
-	c.subAndCallbackByID[sub.ID()] = subAndCallback{
-		cb:  func(p *rtp.Packet) error { return packetsCB([]*rtp.Packet{p}) },
-		sub: sub,
-	}
-
-	// Start the subscription to process calls to sub.Publish
-	sub.Start()
```

Comment:
I moved these calls below as doing so prevents us from needing to do extra cleanup when SubscribeRTP fails